### PR TITLE
Added tmatrix/ttable types, {Matrix}Table.schema.

### PR DESCRIPTION
**Pull request by cseed**
Also fixed a bug, HailType.pretty was missing the initial indent.

A few design questions:
 - Are ttable/tmatrix dtypes?  I'm treating dtype as value types, so no.
 - Unlike HailType, __str__ return a dense but user-readable format, and __repr__ returns code-as-string to recreate the type.
 - I went for the lower case tmatrix and table.  MatrixType and TableType would also be reasonable (and more analogous to HailType).

Also, why is HailType not just Type?  If you're in module hail, I don't see the need for the Hail prefix.

I'll add BaseType, etc. analogous to Scala in another PR.

### CODE REVIEW

**Comment by tpoterba** [comment]
what do you think about keeping these interfaces private while we digest the change and how we want things to look in Python? That way we don't commit to anything right now.

**Comment by cseed** [comment]
Done.  {Matrix}Table._schema is now private.  ttable/tmatrix are exposed, but not documented.

File: hail/python/hail/expr/types.py [hunk 1]
```diff
@@ -835,6 +836,10 @@ def _eq(self, other):
                 and all(self[f] == other[f] for f in self._fields))
 
     def _pretty(self, l, indent, increment):
+        if not self.items():
```
Comment (tpoterba) [inline hail/python/hail/expr/types.py]: prefer `self._fields` for clarity

File: hail/python/hail/matrixtable.py [hunk 2]
```diff
@@ -603,6 +605,14 @@ def __init__(self, mir):
                                     self._entry.items()):
             self._set_field(k, v)
 
+    @property
+    def _schema(self) -> tmatrix:
```
Comment (tpoterba) [inline hail/python/hail/matrixtable.py]: This should be flipped eventually (row type coming from matrix/table type) but that can come later.

Comment (cseed) [reply hail/python/hail/matrixtable.py]: Yes, I'm planning on that.

File: hail/python/hail/expr/types.py (hunk 1 above)
Comment (cseed) [reply hail/python/hail/expr/types.py]: I want the if to match the loop below.  I changed them both to `self._field_types`.

File: hail/python/hail/expr/matrix_type.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    def __eq__(self, other):
+        return (isinstance(other, ttable)
+                and self.global_type == other.global_type
+                and self.col_type == other.col_type
+                and self.col_key == other.col_key
+                and self.row_type == other.row_type
+                and self.row_key == other.row_key
+                and self.entry_type == other.entry_type)
+
+    def __ne__(self, other):
```
Comment (akotlar) [inline hail/python/hail/expr/matrix_type.py]: Is this needed?

https://docs.python.org/3/reference/datamodel.html#object.__eq__
https://docs.python.org/3/reference/datamodel.html#object.__hash__

At least from those sources I don't see a condition where a user defined __ne__ is needed if it's the negation of __eq__, but I could easily be mistaken!

Comment (tpoterba) [reply hail/python/hail/expr/matrix_type.py]: yeah, I don't think we need this here

**Comment by tpoterba** [comment]
> Also, why is HailType not just Type? If you're in module hail, I don't see the need for the Hail prefix.

This is mostly due to a name clash with typing.Type, which we import everywhere.

File: hail/python/hail/expr/matrix_type.py (hunk 3 above)
Comment (cseed) [reply hail/python/hail/expr/matrix_type.py]: Yeah, you're right.  From that link:

 > By default, __ne__() delegates to __eq__() and inverts the result unless it is NotImplemented.

**Comment by cseed** [comment]
Addressed comments.  Also removed `__ne__` from HailType, which I had copied it from.

**Pull request merged**